File: app/src/main/java/net/ducksmanager/whattheduck/PurchaseAdapter.java
package net.ducksmanager.whattheduck;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RadioButton;
import android.widget.TextView;

import net.ducksmanager.persistence.models.dm.Purchase;

import java.util.List;

import androidx.recyclerview.widget.RecyclerView;

public class PurchaseAdapter extends RecyclerView.Adapter<PurchaseAdapter.ViewHolder> {
    static Purchase selectedItem = null;
    private final List<Purchase> items;
    private final Context context;

    static class NoPurchase extends Purchase {
        NoPurchase() {}
    }

    PurchaseAdapter(Context context, List<Purchase> items) {
        this.context = context;
        this.items = items;
    }

    @Override
    public void onBindViewHolder(PurchaseAdapter.ViewHolder purchaseHolder, final int i) {

        Purchase purchase = items.get(i);
        if (purchase != null) {
            purchaseHolder.purchaseCheck.setChecked(purchase == selectedItem);
            boolean isNoPurchase = purchase instanceof NoPurchase;

            purchaseHolder.purchaseDate.setVisibility(!isNoPurchase ? View.VISIBLE : View.GONE);
            purchaseHolder.purchaseName.setVisibility(!isNoPurchase ? View.VISIBLE : View.GONE);
            purchaseHolder.purchaseCheck.setVisibility(View.VISIBLE);
            purchaseHolder.noPurchaseTitle.setVisibility(isNoPurchase ? View.VISIBLE : View.GONE);

            purchaseHolder.purchaseCheck.setContentDescription(purchase.toString());

            if (!isNoPurchase) {
                purchaseHolder.purchaseDate.setText(purchase.getDate());
                purchaseHolder.purchaseName.setText(purchase.getDescription());
            }
        }
    }
    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        LayoutInflater inflater = LayoutInflater.from(context);
        final View view = inflater.inflate(R.layout.row_purchase, viewGroup, false);
        return new ViewHolder(view);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        final RadioButton purchaseCheck;
        final TextView purchaseDate;
        final TextView purchaseName;
        final TextView noPurchaseTitle;

        ViewHolder(View v) {
            super(v);

            purchaseCheck = v.findViewById(R.id.purchasecheck);
            purchaseDate = v.findViewById(R.id.purchasedate);
            purchaseName = v.findViewById(R.id.purchasename);
            noPurchaseTitle = v.findViewById(R.id.nopurchase);

            View.OnClickListener clickListener = v1 -> {
                selectedItem = items.get(getAdapterPosition());
                notifyItemRangeChanged(0, items.size());
            };
            itemView.setOnClickListener(clickListener);
            purchaseCheck.setOnClickListener(clickListener);
        }
    }
}

File: app/src/main/java/net/ducksmanager/whattheduck/PurchaseAdapter.kt
package net.ducksmanager.whattheduck

import android.app.Activity
import android.view.View
import android.widget.TextView

import net.igenius.customcheckbox.CustomCheckBox

import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Comparator
import java.util.Date
import java.util.HashMap
import java.util.Locale

class PurchaseAdapter internal constructor(activity: Activity, items: HashMap<String, Purchase>) : ItemAdapter<PurchaseAdapter.Purchase>(activity, R.layout.row_purchase, ArrayList(items.values)) {

    protected override val onClickListener: View.OnClickListener?
        get() = null

    protected override val comparator: Comparator<Purchase>
        get() = { purchase1, purchase2 ->
            this@PurchaseAdapter.getComparatorText(purchase2)!!
                    .compareTo(this@PurchaseAdapter.getComparatorText(purchase1))
        }

    abstract class Purchase {
        internal var isNoPurchase: Boolean? = null
    }

    internal class SpecialPurchase(noPurchase: Boolean?) : Purchase() {
        init {
            this.isNoPurchase = noPurchase
        }
    }

    class PurchaseWithDate(val id: Int?, internal val purchaseDate: Date, internal val purchaseName: String) : Purchase() {

        init {
            this.isNoPurchase = java.lang.Boolean.FALSE
        }
    }

    override fun getViewHolder(v: View): ViewHolder {
        return ViewHolder(v)
    }

    internal inner class ViewHolder(v: View) : ItemAdapter.ViewHolder(v), View.OnClickListener {
        val purchaseCheck: CustomCheckBox
        val purchaseDate: TextView
        val purchaseTitle: TextView
        val noPurchaseTitle: TextView

        init {

            purchaseCheck = v.findViewById(R.id.purchasecheck)
            purchaseDate = v.findViewById(R.id.purchasedate)
            purchaseTitle = getTitleTextView(v)
            noPurchaseTitle = v.findViewById(R.id.nopurchase)
        }

        override fun onClick(view: View) {
            (view.findViewById<View>(R.id.purchasecheck) as CustomCheckBox).isChecked = true
        }
    }

    override fun onBindViewHolder(holder: ItemAdapter.ViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)

        val purchaseHolder = holder as ViewHolder

        val purchase = getItem(position)
        if (purchase != null) {
            val isNoPurchase = purchase.isNoPurchase

            purchaseHolder.purchaseDate.visibility = if ((!isNoPurchase)!!) View.VISIBLE else View.GONE
            purchaseHolder.purchaseTitle.visibility = if ((!isNoPurchase)!!) View.VISIBLE else View.GONE
            purchaseHolder.purchaseCheck.visibility = View.VISIBLE
            purchaseHolder.noPurchaseTitle.visibility = if (isNoPurchase) View.VISIBLE else View.GONE

            v.minimumHeight = 40
            purchaseHolder.purchaseCheck.contentDescription = purchase.toString()

            purchaseHolder.purchaseCheck.setTag(R.id.check_by_user, java.lang.Boolean.FALSE)
            purchaseHolder.purchaseCheck.isChecked = purchase.toString() == AddIssue.selectedPurchaseHash
            purchaseHolder.purchaseCheck.setTag(R.id.check_by_user, null)

            if ((!isNoPurchase)!!) {
                purchaseHolder.purchaseDate.text = dateFormat.format((purchase as PurchaseWithDate).purchaseDate)
            }
        }
    }

    override fun isHighlighted(i: Purchase): Boolean {
        return false
    }

    override fun getPrefixImageResource(i: Purchase, activity: Activity): Int? {
        return null
    }

    override fun getSuffixImageResource(i: Purchase): Int? {
        return null
    }

    override fun getSuffixText(i: Purchase): String? {
        return null
    }

    override fun getText(p: Purchase): String {
        return (p as? PurchaseWithDate)?.purchaseName ?: ""
    }

    override fun getComparatorText(i: Purchase): String? {
        return if (i.isNoPurchase)
            "^"
        else
            dateFormat.format((i as PurchaseWithDate).purchaseDate)
    }

    override fun getIdentifier(i: Purchase): String {
        return if (i.isNoPurchase)
            "no"
        else
            (i as PurchaseWithDate).id.toString()
    }

    companion object {

        val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
    }
}
